![AIMS Logo](https://www.uni-bielefeld.de/fakultaeten/mathematik/ag/hofmanova/daad-aims/aims-senegal.png)

****

# <center> <b> <span style="color:orange;"> Python Proficiency for Scientific Computing and Data Science (PyPro-SCiDaS)  </span> </b></center>

### <center> <b> <span style="color:green;">An Initiation to Programming using Python (Init2Py) </span> </b></center>
    

****
# <center> <b> <span style="color:blue;">Assignment </span> </b></center>

****
### Instructions:
You should solve these problems using only:
- Standard Python data types (e.g., str, list, dict, int).
- Built-in and user-defined functions.
- Basic flow control (if/else, for loops, while loops).
- Standard library modules such as math
****

## Section 1: Python Fundamentals

### Exercise 1
Compute the following expressions for a natural number $n$ chosen by the user:

a) $$\sum_{\substack{i=0 \\ i \text{ is odd}}}^{n-1} (i + 3)$$


In [1]:
# Your code here
def sum_i(n):
    sum_ = 0
    for i in range (0,n):
        if i%2 ==1:
            sum_=sum_+(i+3)
    return sum_


n = int(input("Enter a natural number n: "))
print(sum_i(n))

Enter a natural number n: 3
4


b) $$\prod_{\substack{i=0 \\ i \text{ is even}}}^{n-1} (i + 3)$$

In [2]:
# Your code here
def prod_i(n):
    prod_ = 1
    for i in range (0,n):
        if i%2 ==0:
            prod_*=(i+3)
    return prod_
n = int(input("Enter a natural number n: "))
print(prod_i(n))


Enter a natural number n: 3
15


### Exercise 2
a) Write python code that computes the sum of the square of the first $n$ natural numbers. One version with a "for loop" and another version with a "while loop".

In [3]:
# Part a: For loop version

n = 3

total = 0

for i in range(1, n + 1):
    total += i**2

print("Sum of squares:", total)


Sum of squares: 14


In [4]:
# Part a: While loop version
n =3
total =0
while n>0:
    total += n**2
    n-=1 
    
print("Sum of squares:", total)


Sum of squares: 14


b) Write python code that prints, for a natural number $M$, the smallest natural number $n$ such that $1^2 + 2^2 + \cdots + n^2 \geq M$. One version with a "for loop" and another version with a "while loop".

In [5]:
# Part b: While loop version
M = 9

total = 0
n = 0

while total < M:
    n += 1
    total += n**2

print("The smallest n is:", n)


The smallest n is: 3


In [6]:
# Part b: For loop version

M = 9
total = 0
n = 0

for i in range(1, M + 1):
    total += i**2
    if total >= M:
        n = i
        break

print("The smallest n is:", n)



The smallest n is: 3


### Exercise 3
Write a python script that asks the user for the values of $x$ and $y$ and uses these inputs to compute the numerical value of the expression:

$$\frac{\sqrt{x^2+y}}{\frac{1}{x} + 2\sin y}$$

In [10]:
# Your code here
import math

x = eval(input("Enter x: "))
y = eval(input("Enter y: "))

num =  math.sqrt(x**2+y)
den = (1/x+2*math.sin(y))
if den != 0:
    ans = num/den
    print("Answer: ",ans)

Enter x: 7
Enter y: 6
Answer:  -17.828520768721546


### Exercise 4

A store charges XOF $12$ per item if fewer than $10$ items are purchased. If the number of items is between $10$ and $99$ (inclusive), the price is XOF $10$ per item. For $100$ items or more, the price drops to XOF $7$ per item.

a) Write a function called `Cost` that takes as input the number of items $n$ and returns the total cost.


In [12]:
# Your code here
def Cost(n):
    if n<10:
        return 12*n
    elif 10<=n<100:
        return 10*n
    else:
        return 7*n


b) Test the function `Cost` for $n = 50 \text{ and } 255$.

In [13]:
# Your code here
n1 = 50 
cost1 = Cost(n1)
print(f"The cost for n = {n1} items is: {cost1} XOF")

n2 = 255
cost2 = Cost(n2)
print(f"The cost for n = {n2} items is: {cost2} XOF")

The cost for n = 50 items is: 500 XOF
The cost for n = 255 items is: 1785 XOF


### Exercise 5
For a fixed real number $x$ and a natural number $n$, we can define recursively $x^n$ using the relations:

$$x^0 = 1 \quad \text{and} \quad x^{n+1} = x \cdot x^n$$

Write a function `power(x,n)` that implements the above recursion.

In [14]:
# Your code here
def power(x,n):
    """
    x: Real number
    n: natural number
    return x^n
    """
    if n==0:
        return 1
    else:
        return x*power(x,n-1)
    
print(f"2^5 = {power(2, 5)}") 

2^5 = 32


### Exercise 6

a) Write two functions: `arithmetic_mean`, which takes two numbers and returns their [**arithmetic mean**](https://en.wikipedia.org/wiki/Arithmetic_mean)
, and `geometric_mean`, which takes two numbers and returns their [**geometric mean**](https://en.wikipedia.org/wiki/Geometric_mean). Then write a function `avggeom` that takes two input numbers and, for 10 iterations, repeatedly computes both their arithmetic mean and geometric mean, using the results from the previous iteration as the new inputs.


In [15]:
# Your code here
import math

 
def arithmetic_mean(m,n):
    """ calculates the mean of m and n"""
    a_mean=(m+n)/2
    return a_mean
    

def geometric_mean(m,n):
    """calculates geometric mean of m and n """
    g_mean=math.sqrt(m*n)
    return g_mean
    

def avggeom(m,n):
    """ input:m,n 
    computes both a_mean and g_mean for 10 iterations"""
    for i in range(10):
        a=arithmetic_mean(m,n)
        g=geometric_mean(m,n)
        m,n = a,g
    return a,g

m, n = 4, 16
a, g = avggeom(m, n)
print(f"After 10 iterations:")
print(f"Arithmetic mean = {a}")
print(f"Geometric mean  = {g}")


After 10 iterations:
Arithmetic mean = 8.97211432115041
Geometric mean  = 8.972114321150412


b) Write a function called `distancefn` that computes the [**Euclidean distance**](https://en.wikipedia.org/wiki/Euclidean_distance) between two points $(x_1, y_1)$ and $(x_2, y_2)$ in the plane. Use it in another program that computes the perimeter and the area of a triangle, given three points in the plane (each given as coordinates).

 


In [16]:
# Your code here
import math

def distancefn(x1, y1, x2, y2):
    """Return the Euclidean distance between two points."""
    return math.sqrt((x2 - x1)**2 + (y2 - y1)**2)

def triangle_properties(x1, y1, x2, y2, x3, y3):
    """Compute the perimeter and area of a triangle given three points."""
    a = distancefn(x1, y1, x2, y2)
    b = distancefn(x2, y2, x3, y3)
    c = distancefn(x3, y3, x1, y1)
    
    perimeter = a + b + c
    
    s = perimeter / 2
    
    area = math.sqrt(s * (s - a) * (s - b) * (s - c))
    
    return perimeter, area


x1, y1 = (0, 0)
x2, y2 = (3, 0)
x3, y3 = (0, 4)

perimeter, area = triangle_properties(x1, y1, x2, y2, x3, y3)

print(f"Perimeter = {perimeter}")
print(f"Area = {area}")



Perimeter = 12.0
Area = 6.0


### Exercise 7

a) Let `L=[2,3,5,7,11,13,17,19,23,29,31,37,41,43,47]`. Use list comprehension to produce a list of the gaps (differences) between consecutive entries in `L`. Then find the maximum gap size and the percentage of gaps that have size $2$.

In [17]:
# Your code here
L=[2,3,5,7,11,13,17,19,23,29,31,37,41,43,47]
gaps=[L[i+1]-L[i]for i in range (len(L)-1)]
max_gap=max(gaps)

print(max_gap)

percentage= 100 * gaps.count(2) / len(gaps)
print(percentage)


6
42.857142857142854


b) i) Using list comprehensions, write a function `ListComp` that takes a list of strings as input and returns three objects:

- a list that consists of the strings with their first characters removed
- a list of the lengths of the strings
- a list that consists of only those strings that are at least 5 characters long


In [18]:
# Your code here
def ListComp(n):
    rem_ch = [y[1:] for y in n if len(y) > 0]
    len_= [len(y) for y in n]
    long= [y for y in n if len(y) >= 5]
    return rem_ch,len_,long

ii) Test your function with the list: `["hello", "world", "python", "programming", "AI", "code"]`

In [19]:
# Your code here
n=["hello", "world", "python", "programming", "AI", "code"]
ListComp(n)

(['ello', 'orld', 'ython', 'rogramming', 'I', 'ode'],
 [5, 5, 6, 11, 2, 4],
 ['hello', 'world', 'python', 'programming'])

### Exercise 8

a) Write a program that repeatedly asks the user to enter product names and prices. Store all of these in a dictionary whose keys are the product names and whose values are the prices. When the user is done entering products and prices, allow them to repeatedly enter a product name and print the corresponding price or a message if the product is not in the dictionary.

In [20]:
# Your code here
products = {}


while True:
    prod_name = input("Name of product : ").strip()
    if prod_name.lower() == 'done':
        break
    price = float(input(f"Enter price of '{prod_name}' price: it is going to be "))
    products[prod_name] = price
while True:
    query = input("Enter product name to check price: ").strip()
    if query.lower() == 'quit':
        break
    if query in products:
        print(f"The price of {query} is {products[query]}")
    else:
        print(f"Sorry, '{query}' is not in the product list.")
        

Name of product : apple
Enter price of 'apple' price: it is going to be 30
Name of product : milk
Enter price of 'milk' price: it is going to be 50
Name of product : orange
Enter price of 'orange' price: it is going to be 20
Name of product : mango
Enter price of 'mango' price: it is going to be 40
Name of product : done
Enter product name to check price: milk
The price of milk is 50.0
Enter product name to check price: banana
Sorry, 'banana' is not in the product list.
Enter product name to check price: quit


b) Using the dictionary created above, allow the user to enter an amount and print out a list of the products whose price is less than that amount.

In [21]:
# Your code here
price = float(input(f"Enter amount: "))
for item in products:
    if price > products[item]:
        print(item)


Enter amount: 40
apple
orange


c) Using the dictionary created above, evaluate the most expensive product.

In [22]:
# Your code here
if products:
    most_expensive = max(products, key=products.get)
    print(f"\nThe most expensive product is '{most_expensive}' at {products[most_expensive]}")
else:
    print("No products in the dictionary.")



The most expensive product is 'milk' at 50.0


d) Using the dictionary created above, calculate the average price of all products.

In [23]:
# Your code here
if products:
    average_price = sum(products.values()) / len(products)
    print(f"The average price is: {average_price}")
else:
    print("No products in the dictionary.")


The average price is: 35.0


e) Using the dictionary created above, update the prices of multiple products at once.

In [24]:
# Your code here
updates = {
    'apple': 85,
    'bread': 75,
    'milk': 60
}

for product, new_price in updates.items():
    if product in products:
        products[product] = new_price
        print(f"Updated '{product}' to {new_price}")
    else:
        print(f"'{product}' not found in the product list")

        
print("\nUpdated product list:")
for product, price in products.items():
    print(f"{product}: {price}")


Updated 'apple' to 85
'bread' not found in the product list
Updated 'milk' to 60

Updated product list:
apple: 85
milk: 60
orange: 20.0
mango: 40.0


### Exercise 9

Use string methods and attributes to solve the following:

a) Write a function called `word_counter` that reads in the file `example.txt` and estimates the total number of words in the file. Assume words are separated by spaces.

In [25]:
# Your code here
def word_counter(filename):
    try:
        with open(filename, 'r') as f:
            text = f.read()
            
        words = text.split()
        return len(words)
    
    except FileNotFoundError:
        print(f"Error: The file not found.")
        return 0

word_counter("example.txt")

10

b) Write a function called `UpLow` that reads the content of `example.txt` and then returns a dictionary containing the total number of uppercase letters and the total number of lowercase letters.

In [26]:
# Your code here
def UpLow(filename):
    counts = {'uppercase': 0, 'lowercase': 0}
    
    try:
        with open(filename, 'r', encoding='utf-8') as f:
            text = f.read()
            
        for char in text:
            if char.isupper():
                counts['uppercase'] += 1
            elif char.islower():
                counts['lowercase'] += 1    
        return counts
    
    except FileNotFoundError:
        print(f"Error: The file not found.")
        return counts

UpLow("example.txt")

{'uppercase': 4, 'lowercase': 28}

c) Write a function called `count_word` that reads in `example.txt` and returns how many times the `target_word` ("the") appears in the text. Take note that the count must be case-insensitive (e.g., "The", "the", "THE" all count as one) and that it must only count whole word matches, not partial matches (e.g., "them" should not count if the target is "the").

In [27]:
# Your code here

def count_word(filename, target_word="the"):
    count = 0
    target_word = target_word.lower() 
    
    try:
        with open(filename, 'r') as f:
            text = f.read()
            
        words = text.split()
        
        for word in words:
            clean_word = ''.join(char for char in word if char.isalnum())
            if clean_word.lower() == target_word:
                count += 1
                
        return count
    
    except FileNotFoundError:
        print(f"Error: The file not found.")
        return 0

count_word("example.txt")

1

d) Write a function called `swap_and_clean` that reads the content of the file `example.txt`. The function should process the text by converting all uppercase letters to lowercase and all lowercase letters to uppercase. It should also remove all characters that are not letters, such as numbers and punctuation, while keeping spaces intact. After performing these modifications, the function should write the resulting text to a new file called `output1.txt`.

In [28]:
# Your code here
def swap_and_clean(input_file='example.txt', output_file='output1.txt'):
    try:
        with open(input_file, 'r') as f:
            text = f.read()
        
        new_text = ''
        for char in text:
            if char.isupper():
                new_text += char.lower()
            elif char.islower():
                new_text += char.upper()
            elif char == ' ':
                new_text += char
        
        with open(output_file, 'w') as f_out:
            f_out.write(new_text)
            
    except FileNotFoundError:
        print(f"Error: The file not found.")

swap_and_clean('example.txt','output1.txt')        

e) Write a function called `palindromes` that reads the content of `example.txt`. The function should first split the text into individual words. For each word, it should be cleaned by converting it to lowercase and removing any leading or trailing punctuation. After cleaning, the function should check whether the word is a [**palindrome**](https://en.wikipedia.org/wiki/Palindrome). A palindrome is a word that reads the same forwards and backwards eg "madam". If a word is a palindrome, the function should print the original, uncleaned version of that word to the console.

In [29]:
# Your code here
import string

def palindromes(filename='example.txt'):
    try:
        with open(filename, 'r') as f:
            text = f.read()
        
        words = text.split()
        
        for word in words:
            clean_word = word.strip(string.punctuation).lower()
            
            if len(clean_word) > 1 and clean_word == clean_word[::-1]:
                print(word)
                
    except FileNotFoundError:
        print(f"Error: The file not found.")

palindromes(filename='example.txt')

madam


f) Write a function called `extract_and_format_numbers` that reads the file `example.txt` line by line. The function should identify any numbers that appear in the text, such as `100000000`. For each number found, it should format the number by inserting commas in the appropriate places (for example, converting `100000000` to `100,000,000`). Each formatted number should then be written on a separate line in a new file called `output2.txt`.

In [30]:
# Your code here
import re

def extract_and_format_numbers(input_file='example.txt', output_file='output2.txt'):
    try:
        formatted_numbers = []
        
        with open(input_file, 'r') as f:
            for line in f:
                numbers = re.findall(r'\d+', line)
                
                for num_str in numbers:
                    num_int = int(num_str)
                    formatted = f"{num_int:,}" 
                    formatted_numbers.append(formatted)
        
        with open(output_file, 'w') as f_out:
            for num in formatted_numbers:
                f_out.write(num + '\n')
        
    
    except FileNotFoundError:
        print(f"Error: The file was not found.")

extract_and_format_numbers('example.txt','output2.txt')

### Exercise 10

The [**digital root**](https://en.wikipedia.org/wiki/Digital_root) of a positive number $n$ is obtained as follows: 
- Add up the digits $n$ to get a new number.
- Add up the digits of that to get another new number. Keep doing this until you get a number that has only one digit. That number is the digital root.

For example, if $n = 45893$, we add up the digits to get $$4 + 5 + 8 + 9 + 3 = 29.$$ We then add up
the digits of $29$ to get $$2 + 9 = 11.$$ We then add up the digits of $11$ to get $$1 + 1 = 2.$$ Since 2 has
only one digit, 2 is our digital root.

Write the Python code of the function `DigRoot` that returns the digital root of an integer $n$ given as input.



In [31]:
# Your code here
def DigRoot(n):
    while n >= 10:  
        n = sum(int(digit) for digit in str(n))  
    return n
n = 45893
print(DigRoot(n))

2


## Section 2: Mathematical Applications

### Exercise 1
Compute the first 30 terms of the sequence defined by $2x_{n+2} - x_{n+1} - 6x_n = 0$ sequence, given any two initial values $x_0$ and $x_1$.

In [32]:
# Your code here
def generate_sequence(x0, x1, n=30):
    
    seq = [x0, x1]
    
    for n in range(2, n):
        x_n = (seq[n-1] + 6 * seq[n-2]) / 2
        seq.append(x_n)
    
    return seq

x0 = 1
x1 = 2
seq = generate_sequence(x0, x1)
print(seq)

[1, 2, 4.0, 8.0, 16.0, 32.0, 64.0, 128.0, 256.0, 512.0, 1024.0, 2048.0, 4096.0, 8192.0, 16384.0, 32768.0, 65536.0, 131072.0, 262144.0, 524288.0, 1048576.0, 2097152.0, 4194304.0, 8388608.0, 16777216.0, 33554432.0, 67108864.0, 134217728.0, 268435456.0, 536870912.0]


### Exercise 2

Let $f$ be a continuous function defined on the closed interval $[a, b]$ such that $f(a)$ and $f(b)$ have opposite signs. By the [**Intermediate Value Theorem**](https://en.wikipedia.org/wiki/Intermediate_value_theorem), there exists a root $x_0 \in (a, b)$ for which $f(x_0) = 0$.

The [**bisection method**](https://fiveable.me/numerical-analysis-i/unit-3/bisection-method-algorithm/study-guide/5LfBknp8yfSe5lse) is an iterative method for approximating the root of such a function. At each step, it divides the interval in half and selects the subinterval where the sign change occurs, repeating this process until the interval width is smaller than a given tolerance ($\epsilon$).

a) Write a function named `bisection(f, a, b, epsilon=1e-2)` that:

- Returns a tuple $(\bar{x}, f(\bar{x}))$, where $\bar{x}$ is an approximation of the root such that
$$
|\bar{x} - x_0| \le \epsilon
$$
if $f(a)\cdot f(b) < 0$.  

- Prints ``Bisection method failed'' if $f(a)\cdot f(b) \ge 0$.


In [33]:
# Your code here
def bisection(f, a, b, epsilon=1e-2):
    
    if f(a) * f(b) >= 0:
        print("Bisection method failed")
        return None

    while (b - a) / 2 > epsilon:
        c = (a + b) / 2 
        fc = f(c)
        
        if f(a) * fc < 0:
            b = c
        else:
            a = c
    
    x_bar = (a + b) / 2
    return x_bar, f(x_bar)




b) Test your function `bisection` for $ f(x) = x^3 -3x - 5$, $a=-10$ and $b=10$

In [34]:
# Your code here
def f(x):
    return x**3 -3*x - 5

root, f_at_root = bisection(f, -10, 10, epsilon=1e-2)
print(f"Approximate root: {root}")
print(f"f(root) = {f_at_root}")


Approximate root: 2.275390625
f(root) = -0.04555877298116684


### Exercise 3
An irrational decimal fraction, [**Champernowne constant**](https://en.wikipedia.org/wiki/Champernowne_constant), is created by concatenating the positive integers: 0.12345678910<font color='red'>1</font>112131415161718192021...

It can be seen that the 12th digit of the fractional part is <font color='red'>1</font>. If $d_n$ represents the nth digit of the fractional part, find the value of the following expression.

$$d_1 × d_{10} × d_{100} × d_{1000} × d_{10000} × d_{100000} × d_{1000000}$$

In [35]:
# Your code here
def champernowne_digit(n):
    length = 1      
    count = 9 
    start = 1
    
    while n > length * count:
        n -= length * count
        length += 1
        count *= 10
        start *= 10
    
    number = start + (n - 1) // length
    digit_index = (n - 1) % length
    return int(str(number)[digit_index])


positions = [1, 10, 100, 1000, 10000, 100000, 1000000]
product = 1
for pos in positions:
    product *= champernowne_digit(pos)
print(product)

210


## Section 3: Advanced Programming

### Exercise 1

a) Write a Python function named `factorial2` that takes a non-negative integer $n$ as input and returns $n!$ (the factorial of $n$).

In [36]:
# Your code here
def factorial2(n):
    result = 1
    for i in range(2, n + 1):
        result *= i
    
    return result

print(factorial2(0))
print(factorial2(5))
print(factorial2(10))


1
120
3628800


b) i) Write a Python function named `QuadEqu` that solves a quadratic equation of the form:
$$ax^2 + bx + c = 0$$

Your function should:
- Take three real numbers $a$, $b$, and $c$ as inputs, with $a$ defaulting to 1.
- Return the real solutions of the equation (if any) in a suitable format.

In [37]:
# Your code here
import math

def QuadEqu(a=1, b=0, c=0):
    if a == 0: 
        if b != 0:
            return (-c / b,)
        else:
            return "No solution" if c != 0 else "Infinite solutions"
    
    delta = b**2 - 4*a*c
    
    if delta > 0:
        x1 = (-b + math.sqrt(delta)) / (2*a)
        x2 = (-b - math.sqrt(delta)) / (2*a)
        return (x1, x2)
    elif delta == 0:
        x = -b / (2*a)
        return (x,)
    else:
        return "No real solutions"

    

ii) Test the function `QuadEqu` for:
  - $a=1, b=2, c=1$
  - $a=2, b=2, c=40$
  - $a=2, b=3, c=-5$

In [38]:
# Your code here
print(QuadEqu(1, 2, 1))
print(QuadEqu(2, 2, 40))
print(QuadEqu(2, 3, -5))

(-1.0,)
No real solutions
(1.0, -2.5)


c) Write a Python program named `fact2nddeg` that repeatedly offers the user a menu with three options:
- Compute a factorial using factorial2
- Solve a quadratic equation using QuadEqu
- Exit the program

The program should:
- Ask the user to select an option.
- Request the necessary input values for the chosen option.
- Call the corresponding function and display the result.
- Repeat the process until the user chooses to exit.

In [40]:
# Your code here
def fact2nddeg():
    while True:
        print("\nMenu:")
        print("1. Compute a factorial")
        print("2. Solve a quadratic equation")
        print("3. Exit")
        
        choice = input("Select an option (1/2/3): ").strip()
        
        if choice == '1':
            # Factorial
            n = int(input("Enter a non-negative integer: "))
            if n < 0:
                print("Please enter a non-negative integer.")
            else:
                print(f"{n}! = {factorial2(n)}")
        
        elif choice == '2':
            # Quadratic equation
            a = float(input("Enter a (default 1): ") or 1)
            b = float(input("Enter b: "))
            c = float(input("Enter c: "))
            result = QuadEqu(a, b, c)
            print(f"Result: {result}")
           
        
        elif choice == '3':
            print("Exiting...")
            break
        
        else:
            print("Invalid choice. Please select 1, 2, or 3.")
fact2nddeg()


Menu:
1. Compute a factorial
2. Solve a quadratic equation
3. Exit
Select an option (1/2/3): 1
Enter a non-negative integer: 3
3! = 6

Menu:
1. Compute a factorial
2. Solve a quadratic equation
3. Exit
Select an option (1/2/3): 2
Enter a (default 1): 1
Enter b: 2
Enter c: 1
Result: (-1.0,)

Menu:
1. Compute a factorial
2. Solve a quadratic equation
3. Exit
Select an option (1/2/3): 3
Exiting...


### Exercise 2
This exercise builds the essential mathematical tools used in public-key cryptography, like [**Rivest–Shamir–Adleman (RSA)**](https://en.wikipedia.org/wiki/RSA_cryptosystem). Security relies on the unique properties of prime numbers and the extreme difficulty of factoring very large numbers.

The Core Principle: Prime Factoring is Hard
   - **Easy**: Multiplying two large prime numbers ($p$ and $q$) to get a huge composite number ($N=p \cdot q$) is very fast for computers.
   - **Hard**: Given only the product $N$, finding the original two prime factors, $p$ and $q$, takes an enormous amount of time. The time required grows exponentially with the size of $N$. This difficulty is what guarantees security.

a) Write a function called `number_of_factors` that takes a positive integer $n$ and returns how many factors the number has together with a list of its factors.

In [41]:
# Your code here
def number_of_factors(n):
    
    factors = []
    
    for i in range(1, int(n**0.5) + 1):
        if n % i == 0:
            factors.append(i)
            if i != n // i:
                factors.append(n // i)
    
    factors.sort()
    return len(factors), factors

n = 28
count, factors = number_of_factors(n)
print(f"{n} has {count} factors: {factors}")


28 has 6 factors: [1, 2, 4, 7, 14, 28]


b) Write a Python function called `is_prime` that takes a positive integer, $n$, as input and returns the boolean True if $n$ is a prime number and False, otherwise.

Note : A prime number (or a prime) is a natural number greater than 1 and that has no positive divisors other than 1 and itself.

In [42]:
# Your code here
def is_prime(n):
    
    if n <= 1:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    
    for i in range(3, int(n**0.5) + 1, 2):
        if n % i == 0:
            return False
    
    return True


print(is_prime(1))
print(is_prime(2))
print(is_prime(17))
print(is_prime(20))


False
True
True
False


c) Cryptographic systems need a large pool of primes. Write a function called `primes(k)` that returns a list of the first k prime numbers. (Set the default value of $k$ to 100).

In [43]:
# Your code here
def primes(k=100):
    prime_list = []
    n = 2
    
    while len(prime_list) < k:
        if is_prime(n):
            prime_list.append(n)
        n += 1
    
    return prime_list


first_10_primes = primes(10)
print(first_10_primes)


[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]


d) Modify the `primes` function to accept an optional argument called start (default value is 2). The function should now return the first $k$ prime numbers that are greater than or equal to start. This simulates searching for primes in a specific high-security range.

In [44]:
# Your code here
def primes(k=100, start=2):

    prime_list = []
    n = max(2, start)
    
    while len(prime_list) < k:
        if is_prime(n):
            prime_list.append(n)
        n += 1
    
    return prime_list

first_10_primes_from_50 = primes(10, start=50)
print(first_10_primes_from_50)

[53, 59, 61, 67, 71, 73, 79, 83, 89, 97]


e) The [**Fundamental theorem of arithmetic**](https://en.wikipedia.org/wiki/Fundamental_theorem_of_arithmetic) states that every integer $n$ is a prime number or can be uniquely written as a product of prime powers (e.g., $12=2^2 \cdot 3^1$. We are interested in finding the exponent of a specific prime in such a factorization. Write a function `V(n, p)` that takes a positive integer $n$ and a prime $p$, and returns the largest integer $r$ such that $p^r$ divides $n$. This value, $r$, is known as the [**p-adic valuation**](https://en.wikipedia.org/wiki/P-adic_valuation).

In [45]:
# Your code here
def V(n, p):
    if n <= 0:
        raise ValueError("n must be a positive integer")
    if not is_prime(p):
        raise ValueError("p must be a prime number")
    
    r = 0
    while n % p == 0:
        n //= p
        r += 1
    return r

print(V(12, 2))
print(V(12, 3))
print(V(25, 5))
print(V(17, 2))


2
1
2
0


f) We would like now to compare the prime-factor structure of two numbers, $n$ and $m$, that is, we would like to measure how different two numbers are in terms of the powers of the primes in their factorizations. Write a function `D(n, m)` that takes two positive integers, $n$ and $m$, and returns a total score $S_1$ calculated as follows:

 $$ S_1 = \sum_{p\; prime} \Bigl| \nu_p(n) - \nu_p(m)\Bigr| $$

Note that `D(14,24)=4`.



In [46]:
# Your code here
def D(n, m):
    if n <= 0 or m <= 0:
        raise ValueError("n and m must be positive integers")
    
    max_val = max(n, m)
    prime_list = primes()
    while prime_list[-1] < max_val:
        prime_list += primes(len(prime_list), prime_list[-1] + 1)
    
    S1 = 0
    for p in prime_list:
        if p > max_val:
            break
        S1 += abs(V(n, p) - V(m, p))
    
    return S1

print(D(14, 24))

4


g) The final problem requires you to analyze the average behavior of the `D(n,m)` function by computing a large, iterated sum.  Hence, write a function `P` that takes a positive integer `N` and returns the sum $S_2$ given by:

$$S_2  = \sum_{n=1}^{N}\sum_{m=1}^{N}D(n,m).$$

Note that $P(10)=210$ and $P(10^2) = 37018$.


In [47]:
# Your code here
def P(N):
    S2 = 0
    for n in range(1, N+1):
        for m in range(n+1, N+1):
            diff = D(n, m)
            S2 += 2 * diff 
    return S2

print(P(10))
print(P(100))

210
37018


### Exercise 3

Imagine a population of bacteria that grows at a rate proportional to its current size, but with limited food supply. Suppose its growth is modeled by

$$\dfrac{dy}{dt}=0.5y$$

with initial population $y(0)=2$.

a) Explain in words what this equation means about how the population changes over time.

**Your explanation here**

This equation means that the population growth rate at any moment is exactly 50% of its current size per unit of time. This means the population is experiencing exponential growth at a continuous relative rate of 0.5 (or 50%).

b) [**Euler’s method**](https://en.wikipedia.org/wiki/Euler_method) is a numerical technique to approximate solutions of ordinary differential equations (ODEs) when an exact solution is hard (or impossible) to find.

Suppose we have an ODE:
$$\dfrac{dy}{dt} = f(t, y), \quad y(t_0)=y_0$$

The idea is:

- Start from an initial point $(t_0, y_0)$
- Move forward in small steps of size $h$
- At each step, estimate the new value of $y$ using the slope $f(t,y)$ at the current point

The update formula is thus given by:
$$y_{n+1} = y_n + h \cdot f(t_n,y_n)$$
$$t_{n+1} = t_n + h$$

The pseudocode is given as follows:

c) i) Write $f(t, y) = 0.5y$ as a lambda function.

In [48]:
# Your code here

f = lambda t, y: 0.5 * y

ii) Now, use Euler’s method with a step size $h=0.1$ to approximate the population at $t=0.3$. Strictly use lists.

In [49]:
# Your code here

t0 = 0
y0 = 2
h = 0.1
n = int(0.3 / h) 


def euler(f, t0, y0, h, n):
    t = [t0]
    y = [y0]

    for i in range(n):
        y_new = y[i] + h * f(t[i], y[i])
        t_new = t[i] + h
        y.append(y_new)
        t.append(t_new)
        
    return y[-1]
    

print(f"Approximate population at t = 0.3: {euler(f, t0, y0, h, n)}")

Approximate population at t = 0.3: 2.205


d) Compare your result at $t=0.3$ with the exact solution $y(t) = 2\exp^{0.5t}$, that is, compute the absolute error.

In [50]:
# Your code here

y_euler = euler(f, t0, y0, h, n)

y_exact = 2 * math.exp(0.5 * 0.3)

abs_error = abs(y_exact - y_euler)

print(f"Euler approximation at t = 0.3: {y_euler}")
print(f"Exact solution at t = 0.3: {y_exact}")
print(f"Absolute error: {abs_error}")

Euler approximation at t = 0.3: 2.205
Exact solution at t = 0.3: 2.323668485456566
Absolute error: 0.11866848545656605


e) Compare your answers for $y(0.3)$ with $h=0.1$ and $h=0.05$. Which one is closer to the exact solution $y(0.3) = 2\exp^{0.15}$ ? Reflect on this.

In [52]:
# Your code here

y_exact = 2 * math.exp(0.5 * 0.3)


t0 = 0
y0 = 2

y_h1 = euler(f, t0, y0, 0.1, int(0.3 / 0.1))


y_h2 = euler(f, t0, y0, 0.05, int(0.3 / 0.05))




error_h1 = abs(y_exact - y_h1)
error_h2 = abs(y_exact - y_h2)

print(f"Exact solution at t = 0.3: {y_exact}\n")

print(f"Euler approximation with h = 0.1: {y_h1}")
print(f"Absolute error (h = 0.1): {error_h1}\n")

print(f"Euler approximation with h = 0.05: {y_h2}")
print(f"Absolute error (h = 0.05): {error_h2}\n")

# Reflection
if error_h2 < error_h1:
    print("Smaller step size (h = 0.05) gives a more accurate result.")
else:
    print("Larger step size (h = 0.1) is somehow more accurate, which is unusual.")


Exact solution at t = 0.3: 2.323668485456566

Euler approximation with h = 0.1: 2.205
Absolute error (h = 0.1): 0.11866848545656605

Euler approximation with h = 0.05: 2.26281642578125
Absolute error (h = 0.05): 0.06085205967531593

Smaller step size (h = 0.05) gives a more accurate result.


### Exercise 4

Determine how many Sundays fell on the first day of a month during the twentieth century, that is, from 1 January 1901 to 31 December 2000.

The following information might be useful (though you may also verify or research independently):

- 1 January 1900 was a Monday.
- The number of days in each month is as follows:
    - September, April, June, and November each have 30 days.
    - All other months have 31 days, except February, which has 28 days in a common year and 29 days in a leap year.
- Leap year rules:
    - A year is a leap year if it is divisible by 4.
    - However, if the year is a century (i.e., divisible by 100), it is only a leap year if it is also divisible by 400.

In [53]:
# Your code here
def count_sundays_on_first():
    month_days = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    
    def is_leap(year):
        return (year % 4 == 0 and (year % 100 != 0 or year % 400 == 0))
    
    weekday = 0 
    sunday_count = 0
    
    for year in range(1900, 2001):
        for month in range(12):
            if year >= 1901 and weekday == 6:
                sunday_count += 1
            
            days = month_days[month]
            if month == 1 and is_leap(year):
                days += 1
            
            weekday = (weekday + days) % 7
    
    return sunday_count

result = count_sundays_on_first()
print(result)

171
